### PR TITLE
Feature/pos end to end

### PR DESCRIPTION
**Pull request by n-hutton**
To test:  (run first ./scripts/ci/install-....sh) even if you have done so before.

Build constellation in build dir.

```./scripts/end_to_end_test/run_end_to_end_test.py ./build ./build/apps/constellation/constellation ./scripts/end_to_end_test/end_to_end_test_pos.yaml```

Note: this command here takes end_to_end_test_pos.yaml which is NOT all of the tests picked up by the CI but useful to only test what you are working on. The CI will run end_to_end_test.yaml which is why the yaml is duplicated

### CODE REVIEW

File: apps/constellation/constellation.cpp [hunk 1]
```diff
@@ -219,7 +220,7 @@ Constellation::Constellation(CertificatePtr certificate, Config config)
   , dag_{GenerateDAG(cfg_.features.IsEnabled("synergetic"), "dag_db_", true, certificate)}
   , dkg_{CreateDkgService(cfg_, certificate->identity().identifier(), muddle_.AsEndpoint())}
   , entropy_{CreateEntropy()}
-  , stake_{CreateStakeManager(cfg_.proof_of_stake, *entropy_)}
+  , stake_{CreateStakeManager(cfg_, cfg_.proof_of_stake, *entropy_)}
```
Comment (ejfitzgerald) [inline apps/constellation/constellation.cpp]: Redundant

File: apps/constellation/constellation.cpp [hunk 2]
```diff
@@ -415,13 +417,14 @@ void Constellation::Run(UriList const &initial_peers, core::WeakRunnable bootstr
   }
 
   // BEFORE the block coordinator starts its state set up special genesis
-  if (cfg_.load_state_file)
+  if (cfg_.proof_of_stake)
```
Comment (ejfitzgerald) [inline apps/constellation/constellation.cpp]: Or load state

File: apps/constellation/constellation.cpp [hunk 3]
```diff
@@ -415,13 +417,14 @@ void Constellation::Run(UriList const &initial_peers, core::WeakRunnable bootstr
...
   {
-    FETCH_LOG_INFO(LOGGING_NAME, "Loading from genesis save file.");
+    FETCH_LOG_INFO(LOGGING_NAME,
+                   "Loading from genesis save file. Location: ", cfg_.stakefile_location);
 
     GenesisFileCreator creator(block_coordinator_, *storage_, stake_.get(), dkg_.get());
 
-    if (cfg_.stakefile_location != "")
+    if (cfg_.stakefile_location == "")
```
Comment (ejfitzgerald) [inline apps/constellation/constellation.cpp]: `.empty()`

File: apps/constellation/health_check_http_module.hpp [hunk 4]
```diff
@@ -49,8 +51,10 @@ class HealthCheckHttpModule : public http::HTTPModule
           bool const chain_execution_complete =
               block_coordinator_.GetLastExecutedBlock() == chain_.GetHeaviestBlockHash();
 
+          bool const dkg_synced = dkg_ ? dkg_->IsSynced() : true;
```
Comment (ejfitzgerald) [inline apps/constellation/health_check_http_module.hpp:54]: Not if a miner

File: apps/constellation/health_check_http_module.hpp [hunk 5]
```diff
@@ -49,8 +51,10 @@ class HealthCheckHttpModule : public http::HTTPModule
...
+
           variant::Variant response            = variant::Variant::Object();
-          response["chain_synced"]             = chain_synced;
+          response["chain_synced"]             = chain_synced && dkg_synced;
```
Comment (ejfitzgerald) [inline apps/constellation/health_check_http_module.hpp]: Sep. field

File: libs/ledger/include/ledger/consensus/stake_manager.hpp [hunk 6]
```diff
@@ -89,6 +99,9 @@ class StakeManager final : public StakeManagerInterface
   StakeSnapshotPtr           current_{};               ///< Most recent snapshot
   BlockIndex                 current_block_index_{0};  ///< Block index of most recent snapshot
   EntropyCache               entropy_cache_{};
+  Mode
+           mode_;  ///< Different modes to operate in - disabling or force enabling the staking mechanism
```
Comment (ejfitzgerald) [inline libs/ledger/include/ledger/consensus/stake_manager.hpp]: Break line

File: libs/ledger/src/consensus/stake_manager.cpp [hunk 7]
```diff
@@ -66,21 +69,53 @@ void StakeManager::UpdateCurrentBlock(Block const &current)
...
 {
-  bool generate{false};
+  FETCH_LOG_INFO(LOGGING_NAME, "Should generate block? Prev: ", previous.body.block_number);
+
+  if (mode_ == Mode::ALWAYS_OFF)
+  {
+    return false;
+  }
+
+  if (mode_ == Mode::ALWAYS_ON)
```
Comment (ejfitzgerald) [inline libs/ledger/src/consensus/stake_manager.cpp]: Delete

File: libs/ledger/src/consensus/stake_manager.cpp [hunk 8]
```diff
@@ -131,7 +166,8 @@ std::size_t StakeManager::GetBlockGenerationWeight(Block const &previous, Addres
     weight = SafeDecrement(weight, 1);
   }
 
-  return weight;
+  // Note: weight must always be non zero
+  return weight + 1;
```
Comment (ejfitzgerald) [inline libs/ledger/src/consensus/stake_manager.cpp]: This is wrong

File: scripts/end_to_end_test/input_files/0.key [hunk 9]
```diff
@@ -0,0 +1 @@
+^�-`�c�g�Y�SʺЛ4g�d��h�aCF�
```
Comment (ejfitzgerald) [inline scripts/end_to_end_test/input_files/0.key]: Not sure we need these

File: scripts/end_to_end_test/input_files/info.txt [hunk 10]
```diff
@@ -0,0 +1,3 @@
+0 : E3vrD/VAAia7wVv+EVMQlXkL2vYGBHW8agX/tUlV9BFGs7VEcB5QgqFvOtAxYyKjCQgYQoRjqx4p71y83bxYYQ==
```
Comment (ejfitzgerald) [inline scripts/end_to_end_test/input_files/info.txt]: Not sure that this should be here

File: scripts/end_to_end_test/run_end_to_end_test.py [hunk 11]
```diff
@@ -174,11 +179,74 @@ def connect_nodes(self, node_connections):
     def start_node(self, index):
         print('Starting Node {}...'.format(index))
 
+        # Useful for seeing the exact command
```
Comment (ejfitzgerald) [inline scripts/end_to_end_test/run_end_to_end_test.py]: Remove

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]

File: scripts/end_to_end_test/run_end_to_end_test.py (hunk 11 above)
Comment (n-hutton) [reply scripts/end_to_end_test/run_end_to_end_test.py]: Done

File: scripts/end_to_end_test/input_files/info.txt (hunk 10 above)
Comment (n-hutton) [reply scripts/end_to_end_test/input_files/info.txt]: done

**Review by ejfitzgerald — APPROVED** [review]

**Pull request merged**